torch/csrc/jit/frontend/ir_emitter.cpp

```diff
@@ -3547,6 +3556,99 @@ struct to_ir {
             ->call(tree->range(), method, named_values, {}, 0));
   }
 
+  Value* emitListLiteral(ListLiteral ll, TypePtr type_hint = nullptr) {
+    auto values = getValues(ll.inputs(), /*maybe_unpack=*/true);
+
+    // determine the element type of the list
+    // if we have a type hint of List[T], use T
+    // if the list is non-empty use type_of(list[0])
+    // otherwise assume it is List[Tensor]
+    TypePtr elem_type = TensorType::get();
+    if (type_hint) {
+      if (type_hint->kind() == TypeKind::ListType) {
+        elem_type = type_hint->expectRef<ListType>().getElementType();
+      } else {
+        // If the type hint was not a List[T] throw an error
+        throw ErrorReport(ll) << "Expected a List type hint but instead got "
+                              << type_hint->repr_str();
+      }
+    } else if (!values.empty()) {
+      std::stringstream ss;
+      auto types = fmap(values, [](const Value* v) { return v->type(); });
+      auto maybe_elem_type = unifyTypeList(types, ss);
+      if (!maybe_elem_type) {
+        throw ErrorReport(ll) << "Lists must contain only a single type\n"
+                              << ss.str();
+      }
+      elem_type = maybe_elem_type.value();
+    }
+
+    for (auto v : values) {
+      std::stringstream ss;
+      if (!v->type()->isSubtypeOfExt(elem_type, &ss)) {
+        throw ErrorReport(ll)
+            << "Lists must contain only a single type, expected: "
+            << elem_type->repr_str() << " but found " << v->type()->repr_str()
+            << " instead.\n"
+            << ss.str();
+      }
+    }
+    Value* result =
+        graph->insertNode(graph->createList(elem_type, values))->output();
+    return result;
+  }
```

Comment:
I had originally started to modify it, since I found a test case in which creating a `ListLiteral` with `Union` or `Optional` fails:

```python
@torch.jit.script
def fn():
    x: Union[List[torch.Tensor], int] = [torch.tensor(3)]
    if isinstance(x, List[torch.Tensor]):
        x.append(torch.tensor(3))
    return x
```

The problem is that the `ListConstruct` is assigned to a Value of type `Union[int, Tensor[]]`. That makes sense because of the `isinstance` check later--if we didn't assign it to be `Union` type, the implied else branch where `x` is an `int` would be an error. However, that fails in the interpreter because the `listConstruct` instruction (interpreter version, note the casing) tries to cast the given element to a `List` when it's typed as a `Union`.

Because we need this out ASAP, I decided to put back the original code and just remove the failing test from the `test_union.py` suite. But I thought it looked a lot better to factor this logic out into a helper function, which is why I kept it this way.

Reply:
If the code is not modified, please leave the refactor in a separate ghstack, it makes it hard to review as is.